05/app.js

```diff
@@ -0,0 +1,113 @@
+class Student {
+  constructor (name, surname) {
+    this.firstName = name,
+    this.surName = surname,
+    this.grades = {}
+  }
+
+  addGrade(subject, grade) {
+    const lowerCaseSubject = subject.toLowerCase();
+    if (!this.grades[lowerCaseSubject]) {
+      const subjects = [
+        "biology",
+        "chemistry",
+        "safety education",
+        "early school education",
+        "physics",
+        "geography",
+        "history",
+        "computer science",
+        "english",
+        "german",
+        "polish",
+        "maths",
+        "music",
+        "visual arts",
+        "nature studies",
+        "technology",
+        "social studies",
+        "physical education",
+      ];
+
+      for (const element of subjects) {
+        if (element === lowerCaseSubject) {
+          this.grades[lowerCaseSubject] = [];
+          break;
+        }
+      }
+
+      if (!this.grades[lowerCaseSubject]) {
+        return console.log("You've provided invalid data.");
+      }
+    }
+
+    if (typeof(grade) === 'number') {
+      this.grades[lowerCaseSubject].push(grade);
+    } else {
+      return console.log("You've provided invalid data.");
+    }
+  }
+
+  getAverageGrade(subject) {
+    if (!this.grades[subject] && subject !== undefined) {
+      return console.log("You've provided invalid data.");
+    }
+
+    if (!subject) {
+      const gradesNumber = Object.values(this.grades).reduce(function (a, b) {
+        if (typeof (a) === 'number' && typeof(b) === 'number') {
+          return a + b;
+        } else if (typeof (a) === 'number') {
+          return a + b.length;
+        } else if (typeof (b) === 'number') {
+          return a.length + b;
+        } else {
+        return a.length + b.length;
+        }
+      });
+      const sum = Object.values(this.grades).reduce(function (a, b) {
+        if (typeof (a) === 'number' && typeof (b) === 'number' ) {
+          return a + b;
+        } else if (typeof (a) === 'number') {
+          return a + b.reduce(function (a, b) {
+            return a + b;
+          })
+        } else if (typeof (b) === 'number') {
+          return a.reduce(function (a, b) {
+            return a + b;
+          }) + b;
+        } else {
+          return (
+            a.reduce(function (a, b) {
+              return a + b;
+            }) +
+            b.reduce(function (a, b) {
+              return a + b;
+            })
+          );
+        }
+      });
```

Comment:
Sporo się tu dzieje. Zdecydowanie bym to uprościł - najpierw wstawił wszystkie liczby do tablicy i obliczył średnią jak w linii 93-98. Może pomocna będzie metoda `concat` lub `flat`

Reply:
flat() zmienia życie. Tego mi brakowało!